### ISSUE TITLE
What is the state of support for AD of BatchNormalization and DropOut layers?

### ISSUE THREAD

**Issue by tranvansang** [pytorch/pytorch#96655]
I have come to this issue from this post.

https://pytorch.org/functorch/stable/notebooks/per_sample_grads.html

## Background

What I am doing requires per-sample gradient (in fact, I migrated from TF, so I do not have much experience with pytorch, but I have a sufficient understanding of NN training).
When reading the post, I could not figure out whether functorch's `vmap` supports AD function of BN and DropOut layers.

In my understanding, these layers are relatively popular.
Because they are not pure functions (different behaviors in training and testing modes), and also not pure (e.g., BN layer accumulates the average across the different forward pass in training mode), which makes me wonder:

## My questions
1. Does functorch's `vmap` support AD function of BN and DropOut layers?
2. If yes, how does it do that?

I tried searching for issues with BatchNormalization or DropOut keywords, but the results were fragmented and I still do not know what is the current state now.
Opacus says that only `EmbeddingBag` is not supported (https://github.com/pytorch/opacus/blob/5aa378ea98df9caf8ca1987ee4d636219267d17e/opacus/grad_sample/functorch.py#L22).
Could anyone tell me the answer?

If possible, updating the docs to clarify the supports for these layers would be great.

Thank you very much.

cc @zou3519 @Chillee @samdow @soumith @kshitij12345 @janeyx99

**Comment by zou3519** [comment]
Dropout is supported, but you likely will need to pass a `randomness='different'` flag to vmap when you use it.

You'll likely run into an error using BatchNorm; https://pytorch.org/functorch/stable/batch_norm.html says the options for when you do. BatchNorm is a bit weird to compute per-sample-gradients for, because it involves computing a statistic across all batches, which breaks the vmap model (that computation for all batches happen in parallel without interaction), so the result you get might be unexpected. Could you let us know what your use case is?

**Comment by tranvansang** [comment]
@zou3519 
Thank you very much for your comment.

I've just recognized that I mistakenly ask the questions in this repo while I was going to ask at https://github.com/pytorch/functorch repo.

I also missed the tutorial you mentioned about BatchNormal in functorch website. The information provided in the tutorial is more than enough.

My questions are now all cleared. That is, to use `vmap` we have to make the function pure (non-side effect).

In my case, I need per-sample gradient in a batch to evaluate the Fisher Information Matrix (FIM), and use FIM to precondition the gradient (kind of KFAC and typical second order training methods).

**Comment by zou3519** [comment]
Thanks for the info! Please feel free to open more issues if you run into problems
